### ISSUE TITLE
GNNexplainer results are constant

### ISSUE THREAD

**Issue by jan0b** [pyg-team/pytorch_geometric#5114]
### 🐛 Describe the bug

# Description
I am trying to explain two models that are trained differently using the GNNExplainer. I use PyG's own implementation. From my point of view the results seem weird, as the returned mask values are always the same independent of the loaded model parameters. Please find the used model checkpoints in my repo: https://github.com/jan0b/checkpoints.git. The files only contain the parameter dict of each model.

# Code
```
import torch
from torch_geometric.nn import GNNExplainer, GCNConv, InstanceNorm, global_mean_pool
from ogb.graphproppred import PygGraphPropPredDataset

class Model2Explain(torch.nn.Module):
    def __init__(self, hidden_channels, dataset, model_type:str):
        super(Model2Explain, self).__init__()
        torch.manual_seed(12345)
        self.model_type = model_type
        self.hidden_channels = hidden_channels
        
        if self.model_type == "GCN":
            self.conv1 = GCNConv(dataset.num_node_features, hidden_channels)
            self.conv2 = GCNConv(hidden_channels, hidden_channels*2)
            self.conv3 = GCNConv(hidden_channels*2, hidden_channels*4)
            self.conv4 = GCNConv(hidden_channels*4, hidden_channels*8)

        self.lin = torch.nn.Sequential(torch.nn.Linear(hidden_channels*4, 512), torch.nn.ReLU(), torch.nn.Linear(512, dataset.num_tasks))
        self.norm = InstanceNorm(1, affine=True)

    def forward(self, x, edge_index, batch):
        x = self.norm(x, batch)
        x = self.conv1(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()
        x = self.conv2(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()
        x = self.conv3(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()

        # 2. Readout layer
        x = global_mean_pool(x, batch)  # [batch_size, hidden_channels]

        # 3. Apply a final classifier
        x = self.lin(x)
        
        return x

def get_dataset():

    dataset = PygGraphPropPredDataset(name = "ogbg-molbace", root = '<>')
    split_idx = dataset.get_idx_split()
    train_set = dataset[split_idx["train"]]
    val_set = dataset[split_idx["valid"]]
    test_set = dataset[split_idx["test"]]

    return train_set, val_set, test_set

root = '<>'

m1 = 'model1.pt'
m2 = 'model2.pt'

_, _, dataset = get_dataset()
sample = dataset[0]

device = torch.device('cpu')

model1 = Model2Explain(dataset=dataset, hidden_channels=256, model_type='GCN').to(device)
params_dict1 = torch.load(os.path.join(root, m1))
model1.load_state_dict(params_dict1)
model1.eval()

model2 = Model2Explain(dataset=dataset, hidden_channels=256, model_type='GCN').to(device)
params_dict2 = torch.load(os.path.join(root, m2))
model2.load_state_dict(params_dict2)
model2.eval()

explainer1 = GNNExplainer(model1, epochs=200, return_type='raw', log=True, device=device)
node_feat_mask11, edge_mask11 = explainer1.explain_graph(x=sample.x, edge_index=sample.edge_index)
node_feat_mask21, edge_mask21 = explainer1.explain_graph(x=sample.x, edge_index=sample.edge_index)
node_feat_mask31, edge_mask31 = explainer1.explain_graph(x=sample.x, edge_index=sample.edge_index)

explainer2 = GNNExplainer(model2, epochs=200, return_type='raw', log=True, device=device)
node_feat_mask12, edge_mask12 = explainer2.explain_graph(x=sample.x, edge_index=sample.edge_index)
node_feat_mask22, edge_mask22 = explainer2.explain_graph(x=sample.x, edge_index=sample.edge_index)
node_feat_mask32, edge_mask32 = explainer2.explain_graph(x=sample.x, edge_index=sample.edge_index)

print('\n')
print(model1(sample.x, sample.edge_index, torch.zeros(sample.x.shape[0], dtype=int, device=device)))
print(model2(sample.x, sample.edge_index, torch.zeros(sample.x.shape[0], dtype=int, device=device)))
print('\n')
print(node_feat_mask11)
print(node_feat_mask12)
print('\n')
print(node_feat_mask21)
print(node_feat_mask22)
print('\n')
print(node_feat_mask31)
print(node_feat_mask32)
print('\n')

print(torch.all(node_feat_mask11 == node_feat_mask12), torch.any(node_feat_mask11 == node_feat_mask12))
print(torch.all(node_feat_mask21 == node_feat_mask22), torch.any(node_feat_mask21 == node_feat_mask22))
print(torch.all(node_feat_mask31 == node_feat_mask32), torch.any(node_feat_mask31 == node_feat_mask32))
print(torch.all(edge_mask11 == edge_mask12), torch.any(edge_mask11 == edge_mask12))
print(torch.all(edge_mask21 == edge_mask22), torch.any(edge_mask21 == edge_mask22))
print(torch.all(edge_mask31 == edge_mask32), torch.any(edge_mask31 == edge_mask32))
```

# Result
This is my result, when the models are initialized, but the **model parameters are not loaded**
```
tensor([[0.0763]], grad_fn=<AddmmBackward0>)
tensor([[0.0763]], grad_fn=<AddmmBackward0>)


tensor([0.1396, 0.1440, 0.1380, 0.1490, 0.1510, 0.1317, 0.1321, 0.1376, 0.1437])
tensor([0.1424, 0.1329, 0.1507, 0.1423, 0.1372, 0.1393, 0.1305, 0.1412, 0.1419])


tensor([0.1367, 0.1253, 0.1454, 0.1487, 0.1351, 0.1343, 0.1610, 0.1464, 0.1457])
tensor([0.1480, 0.1386, 0.1512, 0.1468, 0.1398, 0.1331, 0.1419, 0.1432, 0.1472])


tensor([0.1457, 0.1383, 0.1385, 0.1322, 0.1409, 0.1342, 0.1363, 0.1603, 0.1357])
tensor([0.1406, 0.1451, 0.1261, 0.1391, 0.1511, 0.1390, 0.1497, 0.1274, 0.1393])


tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
```
This is my result, when I **only load the m1 model checkpoint**.
```
tensor([[-3.1716]], grad_fn=<AddmmBackward0>)
tensor([[0.0763]], grad_fn=<AddmmBackward0>)


tensor([0.1396, 0.1440, 0.1380, 0.1490, 0.1510, 0.1317, 0.1321, 0.1376, 0.1437])
tensor([0.1424, 0.1329, 0.1507, 0.1423, 0.1372, 0.1393, 0.1305, 0.1412, 0.1419])


tensor([0.1367, 0.1253, 0.1454, 0.1487, 0.1351, 0.1343, 0.1610, 0.1464, 0.1457])
tensor([0.1480, 0.1386, 0.1512, 0.1468, 0.1398, 0.1331, 0.1419, 0.1432, 0.1472])


tensor([0.1457, 0.1383, 0.1385, 0.1322, 0.1409, 0.1342, 0.1363, 0.1603, 0.1357])
tensor([0.1406, 0.1451, 0.1261, 0.1391, 0.1511, 0.1390, 0.1497, 0.1274, 0.1393])


tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
```
This is my result, when I **load both model checkpoints m1 and m2**.
```
tensor([[-3.1716]], grad_fn=<AddmmBackward0>)
tensor([[2.1335]], grad_fn=<AddmmBackward0>)


tensor([0.1396, 0.1440, 0.1380, 0.1490, 0.1510, 0.1317, 0.1321, 0.1376, 0.1437])
tensor([0.1424, 0.1329, 0.1507, 0.1423, 0.1372, 0.1393, 0.1305, 0.1412, 0.1419])


tensor([0.1367, 0.1253, 0.1454, 0.1487, 0.1351, 0.1343, 0.1610, 0.1464, 0.1457])
tensor([0.1480, 0.1386, 0.1512, 0.1468, 0.1398, 0.1331, 0.1419, 0.1432, 0.1472])


tensor([0.1457, 0.1383, 0.1385, 0.1322, 0.1409, 0.1342, 0.1363, 0.1603, 0.1357])
tensor([0.1406, 0.1451, 0.1261, 0.1391, 0.1511, 0.1390, 0.1497, 0.1274, 0.1393])


tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
tensor(False) tensor(False)
```

I also tried a model with an output shape of (1,2) to make use of a softmax instead of the sigmoid, but it showed the same behaviour.

# Expected Result
The expected result would be that the returned mask values differ if one uses the pretrained model instead of a random initialized model. I would appreciate your help or advise here.

### Environment

* PyG version: 2.0.4 (py38_torch_1.11.0_cu113)
* PyTorch version: 1.11.0 (py3.8_cuda11.3_cudnn8.2.0_0)
* OS: Linux/4.18.0
* Python version: 3.8.13
* CUDA/cuDNN version: cuda11.3_cudnn8.2.0_0
* How you installed PyTorch and PyG (`conda`, `pip`, source): both with conda
* Any other relevant information (*e.g.*, version of `torch-scatter`): ogb 1.3.3

**Comment by rusty1s** [comment]
So the issue is that the masks do not differ for different model weights? This is indeed very weird. Have you checked different hyperparameters for `GNNExplainer`?  Can you debug the gradients of the `edge_mask` during training the explainer to ensure that these are non-zero?

**Comment by jan0b** [comment]
# First
I solved the optimization part of the bug by adapting the _to_log_prob() function of the GNNExplainer (find my patch below) to allow for using the sigmoid. The original code always applies the softmax to the model's logits, but in case of binary classification with only one logit this doesn't work. I would recommend adapting the code to allow for binary classification with only one logit.

```
def _to_log_prob(self, x):
        x = torch.nn.functional.logsigmoid(x) if self.return_type == 'raw' and x.shape[1] == 1 else x
        x = x.log_softmax(dim=-1) if self.return_type == 'raw'  and x.shape[1] != 1 else x
        x = x.log() if self.return_type == 'prob' else x
        return x
```
# Second
The second part of the bug I don't know where the issue lies. If one explains the same input several times it seems the different explainers influence each other, as the importance values maintain the same order. In the code below you can change the order of the blocks, but the output will always remain the same.

```
import torch
from torch_geometric.nn import GNNExplainer, GCNConv, InstanceNorm, global_mean_pool
from ogb.graphproppred import PygGraphPropPredDataset
import os

class MY_GNNExplainer(GNNExplainer):
    def _to_log_prob(self, x):
        x = torch.nn.functional.logsigmoid(x) if self.return_type == 'raw' and x.shape[1] == 1 else x
        x = x.log_softmax(dim=-1) if self.return_type == 'raw'  and x.shape[1] != 1 else x
        x = x.log() if self.return_type == 'prob' else x
        return x

class Model2Explain(torch.nn.Module):
    def __init__(self, hidden_channels, dataset, model_type:str):
        super(Model2Explain, self).__init__()
        torch.manual_seed(12345)
        self.model_type = model_type
        self.hidden_channels = hidden_channels
        
        if self.model_type == "GCN":
            self.conv1 = GCNConv(dataset.num_node_features, hidden_channels)
            self.conv2 = GCNConv(hidden_channels, hidden_channels*2)
            self.conv3 = GCNConv(hidden_channels*2, hidden_channels*4)
            self.conv4 = GCNConv(hidden_channels*4, hidden_channels*8)

        self.lin = torch.nn.Sequential(torch.nn.Linear(hidden_channels*4, 512), torch.nn.ReLU(), torch.nn.Linear(512, dataset.num_tasks))
        self.norm = InstanceNorm(1, affine=True)

    def forward(self, x, edge_index, batch):
        x = self.norm(x, batch)
        x = self.conv1(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()
        x = self.conv2(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()
        x = self.conv3(x, edge_index)
        x = self.norm(x, batch)
        x = x.relu()

        # 2. Readout layer
        x = global_mean_pool(x, batch)  # [batch_size, hidden_channels]

        # 3. Apply a final classifier
        x = self.lin(x)
        
        return x

def get_dataset():

    dataset = PygGraphPropPredDataset(name = "ogbg-molbace", root = '<>)
    split_idx = dataset.get_idx_split()
    train_set = dataset[split_idx["train"]]
    val_set = dataset[split_idx["valid"]]
    test_set = dataset[split_idx["test"]]

    return train_set, val_set, test_set

i=0

root = '<>'

m1 = 'model1.pt'
m2 = 'model2.pt'

_, _, dataset = get_dataset()
data = dataset[i]

device = torch.device('cpu')
data = data.to(device)

model1 = Model2Explain(dataset=dataset, hidden_channels=256, model_type='GCN').to(device)
params_dict1 = torch.load(os.path.join(root, m1))#, map_location=device)
model1.load_state_dict(params_dict1)
model1.eval()

print(model1(data.x, data.edge_index, torch.zeros(data.x.shape[0], dtype=int, device=device)))

#---all block---
explainer2 = MY_GNNExplainer(model1, epochs=200, return_type='raw', log=False)
node_feat_mask21, edge_mask21 = explainer2.explain_graph(x=data.x, edge_index=data.edge_index)
node_feat_mask22, edge_mask22 = explainer2.explain_graph(x=data.x, edge_index=data.edge_index)
node_feat_mask23, edge_mask23 = explainer2.explain_graph(x=data.x, edge_index=data.edge_index)

print(node_feat_mask21)
print(node_feat_mask22)
print(node_feat_mask23)
#---------------

#---for block---
ll = []
for r in range(3):
    explainerr = MY_GNNExplainer(model1, epochs=200, return_type='raw', log=False)
    node_feat_maskrr, edge_maskrr = explainerr.explain_graph(x=data.x, edge_index=data.edge_index)
    ll.append(node_feat_maskrr)

for expl in ll:
    print(expl)
#---------------

#---ind block---
explainer11 = MY_GNNExplainer(model1, epochs=200, return_type='raw', log=False)
node_feat_mask11, edge_mask11 = explainer11.explain_graph(x=data.x, edge_index=data.edge_index)
explainer12 = MY_GNNExplainer(model1, epochs=200, return_type='raw', log=False)
node_feat_mask12, edge_mask12 = explainer12.explain_graph(x=data.x, edge_index=data.edge_index)
explainer13 = MY_GNNExplainer(model1, epochs=200, return_type='raw', log=False)
node_feat_mask13, edge_mask13 = explainer13.explain_graph(x=data.x, edge_index=data.edge_index)

print(node_feat_mask11)
print(node_feat_mask12)
print(node_feat_mask13)
#---------------
```

# Result
The output always looks like this:
```
tensor([[-3.1716]], grad_fn=<AddmmBackward0>)
tensor([0.1399, 0.1422, 0.1380, 0.1489, 0.1510, 0.1317, 0.1325, 0.1374, 0.1436])
tensor([0.1368, 0.1227, 0.1454, 0.1487, 0.1351, 0.1343, 0.1615, 0.1463, 0.1455])
tensor([0.1460, 0.1369, 0.1385, 0.1322, 0.1408, 0.1342, 0.1367, 0.1602, 0.1355])
tensor([0.1427, 0.1298, 0.1508, 0.1423, 0.1372, 0.1393, 0.1309, 0.1412, 0.1416])
tensor([0.1483, 0.1367, 0.1512, 0.1468, 0.1398, 0.1331, 0.1423, 0.1430, 0.1470])
tensor([0.1411, 0.1429, 0.1260, 0.1391, 0.1512, 0.1390, 0.1500, 0.1270, 0.1394])
tensor([0.1516, 0.1469, 0.1398, 0.1347, 0.1458, 0.1381, 0.1390, 0.1422, 0.1277])
tensor([0.1382, 0.1430, 0.1422, 0.1466, 0.1406, 0.1504, 0.1440, 0.1526, 0.1563])
tensor([0.1631, 0.1492, 0.1452, 0.1655, 0.1656, 0.1394, 0.1390, 0.1417, 0.1266])
```
I don't think this is desired or am I wrong?

**Comment by rusty1s** [comment]
Glad that you solved your first issue - and sorry the late reply. Can you clarify why the final results look wrong to you? Do you set a seed somewhere which might explain this?

**Comment by jan0b** [comment]
Yes, I found the issue, it was the seed I set in the Model2Explain initialisation. Thanks for your help. So the only 'real' issue is the __to_log_prob_ function.
